exporter/pulsarexporter/factory.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func (f *pulsarExporterFactory) createTracesExporter(
+	_ context.Context,
+	set component.ExporterCreateSettings,
+	cfg config.Exporter,
+) (component.TracesExporter, error) {
+	oCfg := cfg.(*Config)
+	if oCfg.Topic == "" {
+		oCfg.Topic = defaultTracesTopic
+	}
+	if oCfg.Encoding == "otlp_json" {
+		set.Logger.Info("otlp_json is considered experimental and should not be used in a production environment")
+	}
+	exp, err := newTracesExporter(*oCfg, set, f.tracesMarshalers)
+	if err != nil {
+		return nil, err
+	}
+	return exporterhelper.NewTracesExporter(
+		cfg,
+		set,
+		exp.tracesPusher,
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		// Disable exporterhelper Timeout, because we cannot pass a Context to the Producer,
+		// and will rely on the Pulsar Producer Timeout logic.
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
+		exporterhelper.WithRetry(oCfg.RetrySettings),
+		exporterhelper.WithQueue(oCfg.QueueSettings),
+		exporterhelper.WithShutdown(exp.Close))
+}
+
+func (f *pulsarExporterFactory) createMetricsExporter(
+	_ context.Context,
+	set component.ExporterCreateSettings,
+	cfg config.Exporter,
+) (component.MetricsExporter, error) {
+	oCfg := cfg.(*Config)
+	if oCfg.Topic == "" {
+		oCfg.Topic = defaultMetricsTopic
+	}
+	if oCfg.Encoding == "otlp_json" {
+		set.Logger.Info("otlp_json is considered experimental and should not be used in a production environment")
+	}
+	exp, err := newMetricsExporter(*oCfg, set, f.metricsMarshalers)
+	if err != nil {
+		return nil, err
+	}
+	return exporterhelper.NewMetricsExporter(
+		cfg,
+		set,
+		exp.metricsDataPusher,
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		// Disable exporterhelper Timeout, because we cannot pass a Context to the Producer,
+		// and will rely on the sarama Pulsar Timeout logic.
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
+		exporterhelper.WithRetry(oCfg.RetrySettings),
+		exporterhelper.WithQueue(oCfg.QueueSettings),
+		exporterhelper.WithShutdown(exp.Close))
+}
+
+func (f *pulsarExporterFactory) createLogsExporter(
+	_ context.Context,
+	set component.ExporterCreateSettings,
+	cfg config.Exporter,
+) (component.LogsExporter, error) {
+	oCfg := cfg.(*Config)
+	if oCfg.Topic == "" {
+		oCfg.Topic = defaultLogsTopic
+	}
+	if oCfg.Encoding == "otlp_json" {
+		set.Logger.Info("otlp_json is considered experimental and should not be used in a production environment")
+	}
+	exp, err := newLogsExporter(*oCfg, set, f.logsMarshalers)
+	if err != nil {
+		return nil, err
+	}
+	return exporterhelper.NewLogsExporter(
+		cfg,
+		set,
+		exp.logsDataPusher,
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		// Disable exporterhelper Timeout, because we cannot pass a Context to the Producer,
+		// and will rely on the Pulsar Producer Timeout logic.
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
```

Comment:
This is a bit confusing part. Seems like just a no-op copy/paste artifact, Pulsar Producer Timeout logic isn't being used

Reply:
it follows kafkaexporter pattern, do I need to remove it?

Reply:
I found timeout usage in exporterhelper.common.go:
```
// send implements the requestSender interface
func (ts *timeoutSender) send(req request) error {
	// Intentionally don't overwrite the context inside the request, because in case of retries deadline will not be
	// updated because this deadline most likely is before the next one.
	ctx := req.context()
	if ts.cfg.Timeout > 0 {
		var cancelFunc func()
		ctx, cancelFunc = context.WithTimeout(req.context(), ts.cfg.Timeout)
		defer cancelFunc()
	}
	return req.export(ctx)
}
```
If I set timeout = 0(default value = 5s), it will rely on the Pulsar Producer Timeout logic, and Pulsar Producer SendTimeout default value is 30 seconds.

Reply:
> Pulsar Producer SendTimeout default value is 30 seconds.

Ok. Should use our timeout configuration and pass it to Pulsar Producer?

Reply:
It makes sense, I'll change